csrc/scheduler/matmul_utils.cpp

```diff
@@ -307,32 +309,63 @@ std::string getMatmulCompileTimeRejectReason(Fusion* fusion) {
   // 1. check if there is exactly one MmaOp defined in the fusion
   // 2. check if MmaOp inputs match any of supported inputs layout
   // 3. check if fusion represents expressions that are recognized by matmul
-  //    scheduler
+  // scheduler In addition to the above, it may be the case that there is no mma
+  // op, but a mul-sum pair which may be replaced by a mma op. In that case we
+  // run a pass through the IR to find such a pair of mul-sum, and the
+  // properties of that pair. We then test if the mma op that would replace the
+  // mul-sum pair can be scheduled.
 
   // #1
   auto mma_exprs = ir_utils::getOpsOfType<MmaOp>(fusion);
+  // Initializing the machinery to check if there's a Mul-Sum pair
+  // can be replaced by a Mma Op.
+  mma_utils::CombineMulSum combiner(fusion);
+  std::vector<mma_utils::MmaOrMulSumProperties> mma_from_mul_sums = {};
+  mma_utils::MmaOrMulSumProperties::InputsOutputs mma_inputs_out;
+
   if (mma_exprs.size() != 1) {
-    std::stringstream ss;
-    ss << "Matmul scheduler supports fusions only with a single MMA op, got: "
-       << mma_exprs.size();
-    return ss.str();
+    // If there are no mma ops, check if there's a suitable mul-sum pair
+    // that can be substituted by a mma op.
+    if (mma_exprs.empty()) {
+      mma_from_mul_sums = combiner.generateMulSumCanidates();
+    }
+    // There are more than 1 mma op, or there no mma ops and no suitable mul-sum
+    // pair.
+    if (mma_from_mul_sums.size() != 1) {
+      std::stringstream ss;
+      ss << "Matmul scheduler supports fusions only with a single MMA op, got: "
+         << mma_exprs.size();
+      return ss.str();
+    }
+  }
+
+  if (!mma_exprs.empty()) {
+    mma_inputs_out = {
+        static_cast<TensorView*>(mma_exprs.front()->inA()),
+        static_cast<TensorView*>(mma_exprs.front()->inB()),
+        static_cast<TensorView*>(mma_exprs.front()->out())};
   }
 
   // #2
   {
-    const auto input_layout_opt = mma_utils::getMmaLayout(fusion);
+    const auto input_layout_opt = mma_utils::getMmaLayout(
+        fusion,
+        mma_exprs.empty() ? mma_from_mul_sums.front().insouts : mma_inputs_out);
     if (!input_layout_opt.isValid()) {
       return input_layout_opt.getErrorMsg();
     }
   }
 
   // #3
   {
-    for (auto mma_expr : mma_exprs) {
-      auto support_status = isMatmulFusionDefinitionSupported(fusion, mma_expr);
-      if (!support_status.empty()) {
-        return support_status;
-      }
+    auto support_status = isMatmulFusionDefinitionSupported(
+        fusion,
+        mma_exprs.empty() ? mma_from_mul_sums.front().insouts : mma_inputs_out);
+    if (!support_status.empty()) {
+      return support_status;
+    }
+    if (!mma_from_mul_sums.empty()) {
+      combiner.replaceWithMmaOp();
```

Comment:
I think `getMatmulCompileTimeRejectReason` should just say "yes I will accept the fusion" or "No, I don't want to take this fusion because blabla". It should not modify the fusion.

Reply:
Yeah, I completely agree. 
But, how about the function, MatmulScheduler::canScheduleCompileTime (which invokes `getMatmulCompileTimeRejectReason`)?
We could also add a new member function to MatmulScheduler `MatmulScheduler::combineMulSum` to modify the fusion, though that may go again the mul sum being the default.

Reply:
I think `canScheduleCompileTime` should also be ready-only. Does it make sense to do it at the very beginning of `scheduleMatmul`?

Reply:
I was thinking of:
1. Adding a private flag in `MatmulScheduler` called `combineMulSum` which I can set inside `MatmulScheduler::canScheduleCompileTime`. Then in `MatmulScheduler::canScheduleCompileTime` run the combination pass before calling `MatmulScheduler::canScheduleCompileTime()`.

The only problem that I can think of is, will `computeHeuristics` be called before I call `MatmulScheduler::canScheduleCompileTime()` since `computeHeuristics` searches for a Mmm Op and isn't a member function.

Reply:
The scheduler should convert a mul-sum pattern to `MmaOp` if found. It shouldn't do anything if there's already an `MmaOp` expr. I'm not sure why we would need a flag.